File: genefunk/src/main/java/de/hipp/pnp/genefunk/GeneFunkGenome.java
package de.hipp.pnp.genefunk;

import de.hipp.pnp.api.fivee.abstracts.BaseOrigin;
import de.hipp.pnp.base.fivee.Feature5e;
import jakarta.annotation.Nullable;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import lombok.Data;

import java.util.HashSet;
import java.util.Set;

@Entity
@Data
class GeneFunkGenome extends BaseOrigin {

    @Enumerated(EnumType.ORDINAL)
    GeneFunkGenomeType genomeType = GeneFunkGenomeType.ENGINEERED;

    @Id
    String name;

    @Column(columnDefinition = "TEXT")
    String description;

    @Nullable
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    Set<Feature5e> features = new HashSet<>();

    public void addFeature(Feature5e feature5e) {
        features.add(feature5e);
    }
}

File: genefunk/src/main/kotlin/de/hipp/pnp/genefunk/GeneFunkGenome.kt
package de.hipp.pnp.genefunk

import de.hipp.pnp.api.fivee.abstracts.Species
import de.hipp.pnp.base.fivee.Feature5e

class GeneFunkGenome : Species() {
    var genomeType: GeneFunkGenomeType? = GeneFunkGenomeType.UNKNOWN

    override var name: String? = ""

    override var description: String? = ""

    var features: MutableSet<Feature5e?>? = HashSet<Feature5e?>()

    fun addFeature(feature5e: Feature5e?) {
        val featureSet = this.features ?: HashSet<Feature5e?>().also { this.features = it }
        featureSet.add(feature5e)
    }
}
